platform/stm32f4xx/export/platform/i2c_bus.hpp

```diff
@@ -0,0 +1,648 @@
+                I2C_AcknowledgeConfig(i2c, DISABLE);
+            } else if (m_rx_size == 2) {
+                // two bytes reception is a special case, see RM
+                I2C_AcknowledgeConfig(i2c, DISABLE);
+                I2C_NACKPositionConfig(i2c, I2C_NACKPosition_Next);
+            } else {
+                I2C_AcknowledgeConfig(i2c, ENABLE);
+            }
+        }
+
+        IRQ_manager::unmask(irqn);
+        return;
+    }
+
+    // byte was transmitted
+    if (i2c->SR1 & I2C_FLAG_TXE) {
+        if (m_tx_left > 0) {
+            I2C_SendData(i2c, m_tx[m_tx_size - m_tx_left]);
+            m_tx_left--;
+        } else {
+            // all bytes were transmitted
+            I2C_GenerateSTOP(i2c, ENABLE);
+            m_event_handler(channel::tx, event::tc, 0);
+
+            // tx is done here, set up rx
+            if (m_rx) {
+                m_direction = MASTER_RX;
+                i2c_setup_xfer_irq();
+            } else {
+                m_event_handler(channel::meta, event::tc, 0);
+                I2C_ITConfig(i2c, I2C_IT_EVT | I2C_IT_BUF, DISABLE);
+                return;
+            }
+        }
+    }
+
+    // byte was received
+    if (i2c->SR1 & I2C_FLAG_RXNE) {
+        if (m_rx_left == 1 && m_rx_size == 1) { // single byte case
+            *m_rx = I2C_ReceiveData(i2c);
+            m_rx_left = 0;
+            I2C_GenerateSTOP(i2c, ENABLE);
+        } else if (m_rx_size == 2) { // two byte case
+            // here we need to check BTF, according to RM
+            // if it is not set - it will interrupt
+            if (i2c->SR1 & I2C_FLAG_BTF) {
+                I2C_GenerateSTOP(i2c, ENABLE);
+                m_rx[0] = I2C_ReceiveData(i2c);
+                m_rx[1] = I2C_ReceiveData(i2c);
+                m_rx_left = 0;
+            } else {
+                // disable TXE and RXNE interrupt while waiting for BTF
+                I2C_ITConfig(i2c, I2C_IT_BUF, DISABLE);
+            }
+        } else if (m_rx_size > 2) {
+            // see N > 2 bytes reception in RM
+            if (m_rx_left == 3) { // reception from N-2 to N bytes
+                if (i2c->SR1 & I2C_FLAG_BTF) {
+                    I2C_AcknowledgeConfig(i2c, DISABLE);
+                    // read data N-2
+                    m_rx[m_rx_size - m_rx_left] = I2C_ReceiveData(i2c);
+                    m_rx_left--;
+                } else {
+                    // disable TXE and RXNE interrupt while waiting for BTF
+                    I2C_ITConfig(i2c, I2C_IT_BUF, DISABLE);
+                }
+            } else if (m_rx_left == 2) {
+                // check on BTF according to RM
+                if (i2c->SR1 & I2C_FLAG_BTF) {
+                    I2C_GenerateSTOP(i2c, ENABLE);
+                    // read data N-1
+                    m_rx[m_rx_size - m_rx_left] = I2C_ReceiveData(i2c);
+                    m_rx_left--;
+                    // read data N
+                    m_rx[m_rx_size - m_rx_left] = I2C_ReceiveData(i2c);
```

Comment:
yep, 
m_rx[m_rx_size - m_rx_left] = I2C_ReceiveData(i2c);
m_rx_left--;

can be moved to some read_byte method. Thanks